include/envoy/ssl/handshaker.h

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/network/transport_socket.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/ssl/socket_state.h"
+
+#include "openssl/ssl.h"
+
+namespace Envoy {
+namespace Ssl {
+
+class HandshakerCallbacks {
+public:
+  virtual ~HandshakerCallbacks() = default;
+
+  /**
+   * Called when a handshake is successfully performed.
+   */
+  virtual void onSuccessCb(SSL* ssl) PURE;
+  /**
+   * Called when a handshake fails.
+   */
+  virtual void onFailureCb() PURE;
+};
+
+/*
+ * Interface for a Handshaker which is responsible for owning the
+ * `bssl::UniquePtr<SSL>` and performing handshakes.
+ */
+class Handshaker {
+public:
+  virtual ~Handshaker() = default;
+
+  /**
+   * Do the handshake.
+   *
+   * NB: |state| is a mutable reference.
+   */
+  virtual Network::PostIoAction doHandshake(SocketState& state) PURE;
+
+  /**
+   * Set internal pointers to  Network::TransportSocketCallbacks and
+   * Ssl::HandshakerCallbacks.
+   * Depending on impl, these callbacks can be invoked to access connection
+   * state, raise connection events, etc.
+   */
+  virtual void setCallbacks(Network::TransportSocketCallbacks& callbacks,
+                            Ssl::HandshakerCallbacks& handshaker_callbacks) PURE;
+
+  /*
+   * Access the held SSL object as a ptr. Callsites should handle nullptr
+   * gracefully.
+   */
+  virtual SSL* ssl() PURE;
+};
+
+using HandshakerPtr = std::shared_ptr<Handshaker>;
```

Comment:
nit: HandshakerSharedPtr

Reply:
Use of shared_ptr here worries me.  Need to look more closely to understand the full motivations for it.

Reply:
Done.